src/NHibernate/Type/MetaType.cs

```diff
@@ -42,24 +57,29 @@ public override System.Type ReturnedClass
 
 		public override object NullSafeGet(DbDataReader rs, string[] names, ISessionImplementor session, object owner)
 		{
-			object key = baseType.NullSafeGet(rs, names, session, owner);
-			return key == null ? null : values[key];
+			var raw = _baseType.NullSafeGet(rs, names, session, owner);
+			return raw == null ? null : _values?[raw] ?? raw;
 		}
 
-		public override object NullSafeGet(DbDataReader rs,string name,ISessionImplementor session,object owner)
+		public override object NullSafeGet(DbDataReader rs, string name, ISessionImplementor session, object owner)
 		{
-			object key = baseType.NullSafeGet(rs, name, session, owner);
-			return key == null ? null : values[key];
+			var raw = _baseType.NullSafeGet(rs, name, session, owner);
+			return raw == null ? null : _values?[raw] ?? raw;
 		}
 
 		public override void NullSafeSet(DbCommand st, object value, int index, bool[] settable, ISessionImplementor session)
 		{
 			if (settable[0]) NullSafeSet(st, value, index, session);
 		}
 
-		public override void NullSafeSet(DbCommand st,object value,int index,ISessionImplementor session)
+		public override void NullSafeSet(DbCommand st, object value, int index, ISessionImplementor session)
 		{
-			baseType.NullSafeSet(st, value == null ? null : keys[(string)value], index, session);
+			_baseType.NullSafeSet(
+				st,
+				value == null
+					? null
+					: _keys?[(string) value]?? value,
```

Comment:
Can `_keys[(string) value]` (no question mark) return null? I think there is a latent bug here.

Reply:
That would require the `values` `IDictionary<object, string>` supplied to constructor to contain a `null` key entry, which this interface is not supposed to support.

Conversely, we could think about the same question for ` _values[raw]`. But it cannot contains `null` values either, because otherwise building the reversed `_keys` dictionary in constructor would fail.

Reply:
My previous comment is valid for valid keys and values only. If an invalid key or value is supplied, it will yield null.

The previous code would have yield null in such case, while the new code will yield the invalid value (or invalid key for the Get case).

Supplying an invalid key is easy: just alter the data in the database. For this case, the code behavior change is a possible breaking change. An invalid key value in database with previous code is translated to a null association (for an any association). With the new code, it will cast the invalid key to string (and potentially already fail here) then use it as an entity name (and fail if it is not one).

Supplying an invalid value should also be easy: associate the any association with an entity not having a corresponding mapping in the association. There too, the new code could be consider as a possible breaking change: previously, it would have set the association as `null` in the database, ignoring the "unknown" entity. With the new code, it will attempt to save the entity name as being the meta-type key, and will fail if the underlying type is not compatible with a string.

The old behavior does not look very suitable to me: silently ignoring invalid values may be convenient in some cases, but it also causes errors to be harder to diagnose. Still that is an unrelated change.

Reply:
Well no, I am wrong with the second comment. `IDictionary<object, string>` does not yield `null` in case of missing key but throws, so my first comment holds true for both valid and invalid key or values.